### DISCUSSION TITLE
How can I remove the server-side state from the API when I want to reset the state by using a single store?

### DISCUSSION THREAD

**Discussion by KonceptTW** [pmndrs/zustand#2697] [Q&A] [answered]
``` 
// folder structure
.
└── App/
    ├── pages/
    │   └── card/
    │       ├── pageA.jsx
    │       ├── pageB.jsx
    │       ├── pageC.jsx
    │       └── CreateCard.jsx
    ├── store/
    │   └── useCardStore.js
    └── api/
        └── fetcher.js
```

```js
// cardStore.js

import { create } from 'zustand'

const cardData = {
    title: '',
    ... // other data
}

export const useCardStore = create((set, get) => ({
    ...cardData,
    init: (initialState = {}) => {
        set(() => ({
            ...initialState,
        }))
    },
    flush: () => {
        set(() => ({
            ...cardData
        }))
    }
}))
```

``` JSON
// API Response

{
    "id": "abcd",
    "title": "card title",
    ... // other data
}
```

``` jsx
// Page.jsx
// PageA, PageB, PageC

import { useState, useEffect, useRef } from 'react'
import { useParams } from 'react-router-dom'

import { fetcher } from '../../api/fetcher.js'
import { useCardStore } from '../../store/useCardStore.js'

export function Page() {
    const card = useCardStore()
    const { cardId } = useParams()

    useEffect(() => {
        async function getCard() {
            const cardData = await fetcher(`${import.meta.env.API_ENDPOINT}/card/${cardId}`)
            card.init(cardData)
        }

        if( !card.id ) {
             getCard()
        } else {
            // ... do something else
        }
    }, [])
}
```

``` jsx
// CreateCard.jsx

import { useState, useEffect, useRef } from 'react'

import { useCardStore } from '../../store/useCardStore.js'

export function CreateCard() {
    const card = useCardStore()
    
    useEffect(() => {
        card.flush()
    }, [])
}
```

**Description**

Currently, I have several pages using the same data. I check whether the store data contains an id. If it doesn't, I fetch the data from the API.

The default data doesn't include an id; the id comes from the server. So, when returning to the create card page, I attempt to reset the card store back to its default state.

However, when I console.log the card store, the server-side state still exists. Is there any solution that can truly reset the store back to its default state?

These code snippets are mock implementations. There are other server-side keys involved, not just id.

thanks.

**Scenario**

`PageA -> PageB -> CreateCard`
1. The store doesn't have an id, so data is fetched from the API and initialized in the store.
2. Navigate to Page B, where the card data already exists.
3. Attempt to flush the server-side data to prepare for creating a new card, but the server-side data still persists.

**Comment by dbritto-dev** [comment]
@KonceptTW I encourage you to read our nextjs guide

**Reply by KonceptTW** [reply]
Hi @dbritto-dev , thanks for your reply.

You're referring to the [guide](https://zustand.docs.pmnd.rs/guides/nextjs) that suggests using the Context API to provide the same Zustand store and passing that store to each page, right?

However, my main question is whether there is a way to remove the server-side keys that were initialized in the card store.

I want to use the same store to create a new card, so I need to remove the values that were previously initialized in the store's state.

Or do you mean that when I return to the create page, I should create a new Zustand store specifically for creating a new card?

**Reply by dbritto-dev** [reply]
@KonceptTW my bad I thought was a nextjs app haha

**Reply by dbritto-dev** [reply, marked as answer]
@KonceptTW here you go -> https://github.com/pmndrs/zustand/blob/e49b0d07cd141fdae01ffaeeaf021ccc0775368a/docs/reference/use-store.md#using-dynamic-scoped-non-global-vanilla-stores-in-react

**Reply by dbritto-dev** [reply]
We're working on revamping docs that's part of new docs content -> https://github.com/pmndrs/zustand/pull/2508

**Reply by KonceptTW** [reply]
So, essentially, I would need to generate a new card store each time I create a card, similar to the Factory Pattern?

Updated:
Wait, the snippet already mentions `createCounterStoreFactory`. 🙃

**Reply by KonceptTW** [reply]
> @KonceptTW my bad I thought was a nextjs app haha

My bad, my English wasn't clear and caused some misunderstanding. I need to use GPT to help with my grammar. 😅

**Reply by dbritto-dev** [reply]
@KonceptTW yeah, you can use Factory Pattern to create dynamic isolated stores

**Comment by KonceptTW** [comment]
``` js
// cardStore.js

import { createStore } from 'zustand'
import { CARD_ACTIONS } from './action.js'

const cardData = {
    title: '',
    ... // other data
}

const createCardReducer = (state, action) => {
    switch( action.type ) {
        case CARD_ACTIONS.INIT_CARD:
            const newCard = { ...action.payload }

            // I added a console.log here to check if the value is being received.
            // console.log(newCard)

            ...
            // transform data

            return newCard
        
        default:
            return state
    }
}

export const createCardStore = () => createStore()((set, get) => ({
    ...cardData,
    dispatch: action => set((state) => createCardReducer(state, action))
}))
```

``` jsx
// Page.jsx
// PageA, PageB, PageC

import { useState, useEffect, useRef } from 'react'
import { useParams } from 'react-router-dom'
import { useStore } from 'zustand'

import { fetcher } from '../../api/fetcher.js'
import { createCardStore } from '../../store/createCardStore.js'
import { CARD_ACTIONS } from '../../store/action.js'

export function Page() {
    const card = useStore(createCardStore())
    const { cardId } = useParams()

    useEffect(() => {
        async function getCard() {
            const cardData = await fetcher(`${import.meta.env.API_ENDPOINT}/card/${cardId}`)
            card.dispatch({
                type: CARD_ACTIONS.INIT_CARD,
                payload: cardData
            })
        }

        if( !card.id || card.id !== cardId ) {
             getCard()
        } else {
            // ... do something else
        }
    }, [])
}
```

Hi @dbritto-dev , sorry to bother you again.

Actually, I originally implemented my **cardStore** using the **Flux Pattern**.

but when I switched to using the **Factory Pattern** to create independent **cardStore** instances,
I encountered an issue where the state could not be initialized properly.

I checked that the state was being passed into the reducer, but the **cardStore** state wasn’t updating.

I also replaced **createStore** with a **create** function, applied currying, but still encountered the same issue.

I’ve confirmed that my component is using **useStore** to generate the **cardStore**

I’m wondering if this method prevents updating the state through external values?

If this approach isn’t feasible, do you have any suggestions on how I can achieve the desired functionality?

Thank you!

**Reply by dbritto-dev** [reply]
@KonceptTW yeah, that's wrong you shouldn't use `createStore` + `useStore` like that. The usage of `useStore(createCardStore())` is a bad implementation if you want to create an isolated store you need use `Context` and you make sure that you are creating the store only once but your code can create new stores each time the Page renders so after the init is creating a new store without the data

**Reply by KonceptTW** [reply]
This is what I'm trying to achieve at the moment: When I navigate to the 'create' page,
a temporary create card store is generated using **Zustand**, which is only used while I'm on that page.
Once I leave the page, the store will be garbage collected.

Additionally, I have a global card store that can be accessed across different pages. After creating a card, the data returned from the API is updated in the global card store to be displayed across pages.

The example you provided a couple of days ago used a combination of the **Context API** and **new Map**, which I believe implements the **Memoization Pattern**, correct?
It seems to ensure that the stores created throughout the app remain independent and do not conflict with each other—similar to what **TanStack Query** does, but it's a different approach from what I'm trying to achieve now.

Given my current situation, would it be better to use **useState** or **useReducer** instead?

The reason I opted to use **Zustand** to create this page-specific temporary store is that **Zustand** allows me to combine the **Flux Pattern** with **methods-based** approaches within the same store.

``` js
// e.g.

const card = useCardStore()

card.dispatch({
    type: CARD_ACTIONS.INIT_CARD,
    payload: data
})

card.getCardTitle()
```

**Reply by dbritto-dev** [reply]
@KonceptTW would you mind sharing us a minimal repro on stactkblitz?

**Reply by KonceptTW** [reply]
I have an idea that I'm currently working on. If the issue persists, I'll provide a minimal implementation.

If I succeed, I'll also share my solution.

Thank you!

**Reply by KonceptTW** [reply]
@dbritto-dev 

**removed**

This is my final solution: **globalCard** is used for displaying across pages, while **cardStore** is only created when the user navigates to the '**create**' page and is destroyed when they leave. I'm aware that the destroy method has been deprecated, but I'm unsure if there's an alternative way to destroy it upon leaving the page.

p.s. Please disregard the API logic; it's just a free mock API service available online.

**Reply by dbritto-dev** [reply]
@KonceptTW sorry, but that's a bad implementation you are creating multiple stores without make sure that is only created once

**Reply by KonceptTW** [reply]
I understand your concern. You're worried that if the user frequently switches between pages A, B, and C,
it might repeatedly create the Zustand store, right?
If the user switches pages frequently, you're probably concerned about the possibility of a memory leak.

However, the reason I believe it's not yet necessary to switch to a Memorization Pattern is that,
for one, I've logged the globalCard state across pages, and it hasn't been lost.

Additionally, in my current use case, there will only be a single card of data while navigating through these pages.

Later on, I'll be integrating React Query into this project. But due to the tight development timeline, I've decided to postpone optimization for now.

**Reply by dbritto-dev** [reply]
@KonceptTW btw, I'm working on revamping docs starting from reference: apis, hooks, and middlewares

**Reply by dbritto-dev** [reply]
https://github.com/pmndrs/zustand/pull/2706

**Reply by KonceptTW** [reply]
I found how to unsubscribe in the new documentation.

Thank you for all your effort.
